feat#53/ 동적 스케줄링을 통한 페스티벌 상태 변경 구현

HyeonJun0530 opened the pull request:
## 📄 작업 설명
동적 스케줄링을 통해서 페스티벌 상태 변경을 구현했습니다.
일정 시간 마다 도는게 아니라 그 시간에 스케줄 쓰레드 풀에서 가져와서 실행하게 됩니다.
서버 재시작 시 그 전 CronTrigger가 삭제되기 때문에 다시 등록해주고 이미 지나간 시간에 대해서는 크론이 실행되지 않기 때문에 벌크성 쿼리로 수동 업데이트 해줬습니다 ~

더 궁금한게 있다면 코멘트 달아주세요~~

## 🚨 관련 이슈
closes #53

## 🌈 작업 상황
- 동적 스케줄링을 통한 페스티벌 상태 변경
- 스케줄링 쓰레드 풀 설정
- 서버 재시작 시 벌크 쿼리 실행
- Mocking을 통한 테스트 작성

## 📌 기타

Commit:
config: 스케줄링 설정 작성

- 스케줄링 어노테이션 메인 메서드에 작성
- 스케줄링을 위한 쓰레드 풀 사이즈 설정
- 쓰레드 이름 작성

Commit:
feat: 페스티벌 동적 스케줄링 기능 구현

- CronTrigger를 통해 페스티벌 상태를 동적 스케줄링을 통해 페스티벌 시간에 맞춰 상태를 변경
- 페스티벌 상태를 바꾸어 주는 메서드 추가
- 서버 재시작 시 수동으로 상태를 바꾸어 주는 벌크성 로직 작성

Commit:
test: 페스티벌 동적 스케줄링 테스트 작성

 - 동적 스케줄링 Mock 테스트 작성
 - updateFestivalStatus 메서드 테스트 작성

HyeonJun0530 committed:
Merge branch 'develop' into feat#53

Commit:
test: admin 추가에 따른 테스트 수정

- 테스트 데이터에 admin 추가

minji-park-eng commented on src/main/java/com/wootecam/festivals/domain/festival/entity/Festival.java:
```diff
@@ -118,4 +118,8 @@ private void validateTimeRange() {
     }
+
+    public void updateFestivalStatus(FestivalStatus newStatus) {
+        this.festivalStatus = newStatus;
```
p2: 현재 진행 상태 기반으로 유효성 검증 로직이 필요하지 않을까요?

HyeonJun0530 replied:
추후 FestivalStatus 분리에서 해결 하겠습니다~

minji-park-eng commented on src/main/java/com/wootecam/festivals/domain/festival/repository/FestivalRepository.java:
```diff
@@ -26,4 +27,12 @@ List<Festival> findUpcomingFestivalsBeforeCursor(@Param("startTime") LocalDateTi
                                                      Pageable pageable
     );
+
+    @Modifying
+    @Query("UPDATE Festival f SET f.festivalStatus = 'COMPLETED' WHERE f.startTime <= :now AND f.endTime >= :now")
+    void bulkUpdateCompletedFestivals(LocalDateTime now);
+
+    @Modifying
+    @Query("UPDATE Festival f SET f.festivalStatus = 'ONGOING' WHERE f.startTime <= :now AND f.endTime >= :now")
+    void bulkUpdateOngoingFestivals(LocalDateTime now);
```
p2: festivalStatus를 변수로 두면 한 쿼리로 합칠 수 있을 것 같아요~! :)

HyeonJun0530 replied:
감사합니다~ 해결했어요~

minji-park-eng commented on src/main/java/com/wootecam/festivals/domain/festival/service/FestivalSchedulerService.java:
```diff
@@ -0,0 +1,101 @@
+        this.festivalRepository = festivalRepository;
+    }
+
+
```
p3: 개행 삭제 부탁드립니다!

bellringstar commented on src/main/java/com/wootecam/festivals/domain/festival/service/FestivalStatusUpdateService.java:
```diff
@@ -0,0 +1,21 @@
+
+    @Transactional
+    public void updateFestivalStatus(Long festivalId, FestivalStatus newStatus) {
+        Festival festival = festivalRepository.findById(festivalId).orElseThrow();
```
Throw에서 명시적인 예외를 던져주는게 어떨까요? 페스티벌 에러 코드에 관련 에러가 있을 것 같습니다.

bellringstar reviewed (commented):
👍LGTM

minji-park-eng reviewed (commented):
고생하셨습니다! 리뷰 확인 부탁드려요~! :)

Commit:
refactor: 서버 재시작 스케줄링 리팩터링

- findAll() -> 이미 닫힌 페스티벌을 제외하고 가져오게 변경
- 컨벤션에 맞게 줄바꿈
- 업데이트 벌크 쿼리 2개를 1개로 수정
- 테스트 코드 수정

Commit:
refactor: 원자적 처리를 위해 Transactional 추가

Commit:
refactor: FestivalStatusUpdateService 에러 코드 던지도록 수정

HyeonJun0530 committed:
Merge branch 'develop' into feat#53

minji-park-eng reviewed (approved):
고생하셨습니다!! :)
